v2.0.2 scan job fails on every consumer PR — `Checkout shared-workflows scripts` step resolves `github.workflow_sha` to caller context

## What

In `dependency-cooldown.yml@v2.0.2`, the new `Checkout shared-workflows scripts` step (added in #28) fails with `fatal: remote error: upload-pack: not our ref <SHA>` on **every** consumer PR — bot or human — making v2.0.2 functionally unusable as a drop-in replacement for v2.0.1.

**Observed on [j7an/nexus-mcp#162](https://github.com/j7an/nexus-mcp/pull/162) on 2026-04-13** (the first consumer PR after bumping to v2.0.2). Failing run: [run 24324579821](https://github.com/j7an/nexus-mcp/actions/runs/24324579821).

The failure happens at the `actions/checkout` step, before the `Set initial status` step that would normally short-circuit non-Dependabot PRs to `success`. Relevant log lines:

```
##[group]Run actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd
##[error]fatal: remote error: upload-pack: not our ref c02bda1f213ba942269a22df61d1c18eb9ba2902
The process '/usr/bin/git' failed with exit code 128
##[error]The process '/usr/bin/git' failed with exit code 128
```

The SHA in the error — `c02bda1f213ba942269a22df61d1c18eb9ba2902` — is **not** the v2.0.2 commit (`16b94fd3461af6e9fe6e6174d805c7879fb39bb4`). It is **nexus-mcp PR #162's synthetic `merge_commit_sha`**, verified via two independent queries:

```bash
$ gh api repos/j7an/nexus-mcp/pulls/162 --jq '.merge_commit_sha'
c02bda1f213ba942269a22df61d1c18eb9ba2902

$ gh api repos/j7an/shared-workflows/commits/c02bda1f213ba942269a22df61d1c18eb9ba2902
{
  \"message\": \"No commit found for SHA: c02bda1f213ba942269a22df61d1c18eb9ba2902\",
  \"status\": \"422\"
}
```

The step asked GitHub to fetch nexus-mcp's PR merge ref from `j7an/shared-workflows`, which returned 422 because that SHA doesn't exist in this repo (and never will — it's an ephemeral merge ref for an unrelated PR in a different repo).

Severity assessment: **this is strictly worse than #25/#26/#27 combined.** Those bugs fired only on specific edge cases (rebase-into-sub-cooldown, dirty-then-clean re-scan, YAML list-marker shape). This bug fires deterministically on every single consumer PR. v2.0.2 cannot be used by any consumer until it's fixed.

## Why

Two compounding root causes in v2.0.2's new checkout step:

### Root cause 1 — `github.workflow_sha` is caller-context, not reusable-workflow-context

The step's configuration (from `dependency-cooldown.yml@v2.0.2` lines 31–38):

```yaml
- name: Checkout shared-workflows scripts
  uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
  with:
    repository: j7an/shared-workflows
    ref: \${{ github.workflow_sha }}
    path: shared-workflows
    sparse-checkout: scripts
    persist-credentials: false
```

The intent is visibly to check out the exact same commit of `shared-workflows` that the caller pinned in their `uses:` line — a clean time-of-check/time-of-use property. The chosen mechanism is `github.workflow_sha`, which the author of #28 likely expected to resolve to the reusable workflow's own commit SHA.

It does not. Per GitHub's [Variables documentation](https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables), `GITHUB_WORKFLOW_SHA` / `github.workflow_sha` is:

> The commit SHA that triggered the workflow. The value of this commit SHA depends on the event that triggered the workflow.

In a reusable-workflow-called-from-`pull_request`-event context, this resolves to the caller's event SHA — which for a `pull_request` trigger is the synthetic `merge_commit_sha` that GitHub creates for every PR. That SHA exists in the caller's repo, not in `j7an/shared-workflows`, so `actions/checkout` deterministically fails.

This is not a \"flaky CI\" bug or a timing issue. It is a structural impossibility: the SHA being requested can never exist in the target repository.

### Root cause 2 — the checkout step is unguarded before the non-bot short-circuit

Even if the SHA resolution were fixed, the checkout step runs **before** the `Set initial status` step that short-circuits non-Dependabot PRs to `success`. The step sequence in v2.0.2's `jobs.scan.steps` is:

1. `Harden runner`
2. `Checkout shared-workflows scripts` ← unguarded, runs for all PRs
3. `Set initial status` ← emits `skip=true` for non-bot PRs
4. `Scan and report` — has `if: steps.gate.outputs.skip == 'false'` ← correctly gated

Steps 2 is missing the `if:` guard that step 4 has. Non-bot PRs that would otherwise short-circuit still pay the cost of a cross-repo network checkout, and hit this failure first.

## How

The fix needs to address root cause 1 (required) and should also address root cause 2 (cheap efficiency win).

### Root cause 1 — three candidate fixes

**Option A — Inline the scripts back into the workflow YAML (recommended).**

Reverse the extraction from #28 for the production code path only. Inline `extract-deps.sh` and `check-release-age.sh` contents back into the `run:` blocks of `dependency-cooldown.yml`. Keep the standalone `scripts/extract-deps.sh` and `scripts/check-release-age.sh` files and the bats test suite unchanged — they continue to run in CI against the shared-workflows source tree on shared-workflows' own PRs (the test feedback loop is preserved).

The cost is one acknowledged duplication: the runtime bash in `dependency-cooldown.yml` is physically a copy of the bash in `scripts/*.sh`. This duplication is real but manageable: (a) shared-workflows has its own CI gate that lints the inline bash, (b) the bats tests catch behavior regressions against the standalone scripts, (c) any drift between the two copies shows up in PR diffs that touch either file. A pre-commit check or actionlint rule could enforce sync automatically, but is not required for correctness.

This option restores v2.0.1's self-contained property: the reusable workflow carries all its own code inline and has no runtime dependency on fetching its own source tree.

**Option B — Pass the SHA as a caller-provided input.**

Add a required `workflow_sha` input. Callers must set it to match their `uses:` pin:

```yaml
uses: j7an/shared-workflows/.github/workflows/dependency-cooldown.yml@16b94fd3... # v2.0.2
with:
  workflow_sha: 16b94fd3461af6e9fe6e6174d805c7879fb39bb4
  cooldown_days: 7
```

The workflow then uses `ref: \${{ inputs.workflow_sha }}` in the checkout step. This works, but forces every caller to specify the SHA in two places (in `uses:` and in `workflow_sha:`), and any mismatch silently runs scripts from a different version than the pinned workflow YAML expects. Dependabot's updater would also need to be taught to keep both values in sync, which is not a standard pattern.

Not recommended because of the caller-side footgun, but viable as a forward-compatible evolution if a later release also wants to allow callers to pin scripts independently.

**Option C — Parse `github.workflow_ref` to extract the reusable workflow's SHA.**

`github.workflow_ref` is documented to have the shape `{owner}/{repo}/{path}/{filename}@{ref}`. Add a step that parses the ref:

```yaml
- name: Resolve reusable workflow SHA
  id: resolve
  run: |
    ref=\"\${{ github.workflow_ref }}\"
    sha=\"\${ref##*@}\"
    echo \"sha=\$sha\" >> \"\$GITHUB_OUTPUT\"
- name: Checkout shared-workflows scripts
  uses: actions/checkout@...
  with:
    repository: j7an/shared-workflows
    ref: \${{ steps.resolve.outputs.sha }}
    ...
```

This depends on an empirical question: does `github.workflow_ref` resolve to the reusable workflow's ref or the caller's ref in a reusable-workflow context? GitHub's own documentation is inconsistent across pages on this point, and the behavior has changed historically. **This option must be validated by a test PR before adoption.** If it works, it is the most architecturally clean fix (preserves extraction, preserves time-of-check/time-of-use pinning, no caller-side changes needed). If it doesn't, fall back to Option A.

### Root cause 2 — always apply

Regardless of which fix from root cause 1 is chosen, move the `Set initial status` step **before** the checkout, and guard the checkout with `if: steps.gate.outputs.skip == 'false'`. This ensures non-bot PRs skip the network round-trip entirely and never touch the checkout code path. The step reordering is:

1. `Harden runner`
2. `Set initial status` (emits `skip=true|false`)
3. `Checkout shared-workflows scripts` — **add** `if: steps.gate.outputs.skip == 'false'`
4. `Scan and report` — existing `if: steps.gate.outputs.skip == 'false'`

This also closes a small defense-in-depth hole: non-bot PRs currently cause a cross-repo git clone that runs *even though* the scan is about to skip. With the guard, that's avoided.

### Recommendation

**Ship Option A + root-cause-2 reordering as v2.0.3 immediately.** This is the smallest, most defensible emergency fix — it restores the pre-#28 runtime model (inline scripts) while keeping all of #28's test infrastructure and all three regression fixes for #25/#26/#27. Option C can be explored as a v2.1 architectural follow-up if and when `github.workflow_ref` semantics are confirmed.

Do not let perfect be the enemy of good here: v2.0.2 is currently unshipped in the sense that nothing on GitHub can actually run it successfully. Getting a working v2.0.3 out is more important than choosing the nicest long-term architecture.

## Acceptance criteria

A v2.0.3 release closes this issue iff **all** of the following hold:

1. **Consumer PR passes end-to-end on a non-Dependabot (human) author.**
   A test PR on a fresh branch in a consumer repo (e.g. re-targeting [j7an/nexus-mcp#162](https://github.com/j7an/nexus-mcp/pull/162) to v2.0.3), authored by a non-`dependabot[bot]` user, must produce a `dependency-cooldown / gate` status of `success` with description `Non-bot PR — no cool-down required`. No step in the `scan` job may fail or error out.

2. **Consumer PR passes end-to-end on a `dependabot[bot]` author.**
   The next scheduled Dependabot PR in a consumer repo (or a synthetic one created for the purpose) must complete the scan step successfully and produce a scan comment with the expected \"Packages scanned\" and \"Release Age\" sections. The `dependency-cooldown / gate` status must end at `success` (for a clean, aged bundle) or `pending` (for a cooldown-pending bundle), never `error`.

3. **Checkout step is gated.** Inspection of `dependency-cooldown.yml@v2.0.3` shows the `Checkout shared-workflows scripts` step (if still present) has `if: steps.gate.outputs.skip == 'false'` and runs **after** `Set initial status`. If Option A is chosen and the step is removed entirely, this criterion is trivially satisfied.

4. **No regression of #25/#26/#27.** The existing bats test suite (`tests/extract-deps.bats` and `tests/check-release-age.bats`) continues to pass in CI. The fixture at `tests/fixtures/extract-deps/nexus-mcp-160.diff` remains the regression anchor for #27.

5. **Release notes call out the emergency fix explicitly.** v2.0.3's release notes and its merged PR's body state that v2.0.2 was unshippable due to this issue, link to this issue, and identify which fix option was taken (A, B, or C). Consumers currently pinned to v2.0.2 need a clear signal that they must bump and are not simply taking a cosmetic patch.

6. **A forward-compatibility note is added to the README** under a \"Known caller-side constraints\" section describing which GitHub context variables this workflow does and does not rely on, so future reusable-workflow-internal refactors don't reintroduce the same class of bug.

---

Fixes the blocker on [j7an/nexus-mcp#162](https://github.com/j7an/nexus-mcp/pull/162).

Closed-by SHA (pre-empting #28's pattern): whichever PR closes this should also add a `tests/fixtures/` entry capturing a minimal `pull_request`-triggered reusable workflow call that exercises the fix, so the next refactor can't silently reintroduce the same bug.

--- a/.github/workflows/ci-scripts.yml
+++ b/.github/workflows/ci-scripts.yml
@@ -28,3 +28,17 @@ jobs:
 
       - name: Run bats
         run: bats tests/
+
+  inline-sync:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Harden runner
+        uses: step-security/harden-runner@fa2e9d605c4eeb9fcad4c99c224cee0c6c7f3594 # v2.16.0
+        with:
+          egress-policy: audit
+
+      - name: Checkout
+        uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
+
+      - name: Verify inline bash matches scripts/*.sh
+        run: ./scripts/check-inline-sync.sh

--- a/.github/workflows/dependency-cooldown.yml
+++ b/.github/workflows/dependency-cooldown.yml
@@ -72,6 +72,95 @@ jobs:
           COOLDOWN_DAYS: ${{ inputs.cooldown_days }}
           FAIL_ON_COOLDOWN: ${{ inputs.fail_on_cooldown }}
         run: |
+          # --- BEGIN inline:scripts/extract-deps.sh ---
+          extract_deps() (
+          # extract-deps.sh — parse unified diff on stdin, emit dependency TSV on stdout
+          #
+          # Output: <name>\t<version>\t<ecosystem>  where ecosystem ∈ {actions, pypi}
+          # Exit:   0 on success (possibly zero rows), 2 on malformed input
+          #
+          # Handles BOTH GitHub Actions shapes observed in real PR diffs:
+          #   +        uses: owner/repo@sha # vX.Y.Z        (no list marker)
+          #   +      - uses: owner/repo@sha # vX.Y.Z        (YAML list marker)
+          #
+          # The missing list-marker support in the v2.0.1 regex `^\+\s+uses:` is the
+          # root cause of issue #27 (astral-sh/setup-uv silently dropped from
+          # nexus-mcp#160's cooldown scan).
+
+          set -euo pipefail
+
+          # Dedup sentinel: newline-delimited list of "ecosystem:name" keys.
+          # Using a plain string (not `declare -A`) for bash 3.2 compatibility, since
+          # macOS's system bash is still 3.2 and the bats test invokes `bash` directly.
+          seen=$'\n'
+          rows=()
+
+          input=$(cat)
+
+          # Empty input → exit 0 with no output
+          if [ -z "$input" ]; then
+            exit 0
+          fi
+
+          # Malformed input detection: if non-empty and has none of the unified-diff
+          # markers, reject with exit 2.
+          if ! printf '%s\n' "$input" | grep -qE '^(\+\+\+|---|@@|diff --git)'; then
+            echo "extract-deps.sh: input is not a unified diff" >&2
+            exit 2
+          fi
+
+          while IFS= read -r line; do
+            # Strip CR for CRLF-encoded diffs
+            line="${line%$'\r'}"
+
+            # Skip diff file headers (+++ b/path)
+            [[ "$line" == +++* ]] && continue
+
+            # --- GitHub Actions parser ---
+            # Matches both "+ uses: foo@..." and "+ - uses: foo@..."
+            if [[ "$line" =~ ^\+[[:space:]]+(-[[:space:]]+)?uses:[[:space:]]+([^[:space:]@]+)@[^[:space:]]+(.*)$ ]]; then
+              name="${BASH_REMATCH[2]}"
+              rest="${BASH_REMATCH[3]}"
+
+              # Skip local and docker actions
+              [[ "$name" == ./* ]] && continue
+              [[ "$name" == docker://* ]] && continue
+
+              version=""
+              # Capture version from trailing comment: " # v1.2.3" or " # 1.2.3"
+              # The `v?` is OUTSIDE the capture group so we strip the leading v.
+              if [[ "$rest" =~ \#[[:space:]]*v?([0-9][0-9.]*) ]]; then
+                version="${BASH_REMATCH[1]}"
+              fi
+
+              key="actions:$name"
+              case "$seen" in *$'\n'"$key"$'\n'*) continue ;; esac
+              seen="${seen}${key}"$'\n'
+              rows+=("$name"$'\t'"$version"$'\t'"actions")
+              continue
+            fi
+
+            # --- Python deps parser ---
+            # Skip Python-style comment lines first
+            [[ "$line" =~ ^\+[[:space:]]*# ]] && continue
+
+            if [[ "$line" =~ ^\+[[:space:]]*([a-zA-Z][a-zA-Z0-9_.\-]*)[[:space:]]*(==|\>=|\<=|~=|\!=|\>|\<)[[:space:]]*([0-9][0-9a-zA-Z.\-]*) ]]; then
+              name="${BASH_REMATCH[1]}"
+              version="${BASH_REMATCH[3]}"
+              key="pypi:$name"
+              case "$seen" in *$'\n'"$key"$'\n'*) continue ;; esac
+              seen="${seen}${key}"$'\n'
+              rows+=("$name"$'\t'"$version"$'\t'"pypi")
+              continue
+            fi
+          done <<< "$input"
+
+          if [ ${#rows[@]} -gt 0 ]; then
+            printf '%s\n' "${rows[@]}" | sort -t$'\t' -k1,1
+          fi
+          )
+          # --- END inline:scripts/extract-deps.sh ---
+
           DIFF=$(gh pr diff "$PR_NUMBER" --repo "$GH_REPO")
 
           DEPS=""
@@ -87,7 +176,7 @@ jobs:
           PR_BODY=$(gh pr view "$PR_NUMBER" --repo "$GH_REPO" --json body --jq '.body // ""')
 
           # --- Extract dependencies via standalone script (Task 8) ---
-          DEPS_TSV=$(echo "$DIFF" | bash "${GITHUB_WORKSPACE}/shared-workflows/scripts/extract-deps.sh")
+          DEPS_TSV=$(echo "$DIFF" | extract_deps)
 
           # Populate ACTIONS, PY_DEPS, ACTION_VERSIONS, PY_VERSIONS from DEPS_TSV
           # so the existing advisory-scan loops below can consume them unchanged.

--- a/.github/workflows/dependency-cooldown.yml
+++ b/.github/workflows/dependency-cooldown.yml
@@ -382,6 +382,10 @@ jobs:
           COOLDOWN_FAILURES=0
           AGE_TSV=""
           if [ "$COOLDOWN_DAYS" -gt 0 ]; then
+            # check_release_age must be invoked via command substitution so
+            # bash's parent -e is dropped — this preserves per-row error
+            # tolerance (404s, parse failures, etc. emit error verdicts
+            # instead of killing the step). Do not convert to a direct call.
             AGE_TSV=$(echo "$DEPS_TSV" | check_release_age)
             COOLDOWN_FAILURES=$(echo "$AGE_TSV" | awk -F'\t' '$6=="fail" || $6=="error"' | wc -l | tr -d ' ')
           fi

--- a/.github/workflows/dependency-cooldown.yml
+++ b/.github/workflows/dependency-cooldown.yml
@@ -161,6 +161,156 @@ jobs:
           )
           # --- END inline:scripts/extract-deps.sh ---
 
+          # --- BEGIN inline:scripts/check-release-age.sh ---
+          check_release_age() (
+          # check-release-age.sh — read dep TSV on stdin, emit verdict TSV on stdout
+          #
+          # Schema:
+          #   in:  <name>\t<version>\t<ecosystem>
+          #   out: <name>\t<version>\t<ecosystem>\t<published_iso>\t<age_days>\t<verdict>\t<reason>
+          #
+          # Verdicts: pass | fail | error
+          # Exit: always 0; failures are per-row.
+          #
+          # Bash 3.2 compatible (macOS system bash).
+
+          set -uo pipefail
+
+          : "${COOLDOWN_DAYS:?COOLDOWN_DAYS env var is required}"
+          : "${NOW_EPOCH:=$(date +%s)}"
+
+          # iso_to_epoch <iso> — print unix epoch on stdout, return 1 on parse failure.
+          # Handles both GitHub ("2026-03-29T12:00:00Z") and PyPI ("2026-03-29T12:00:00")
+          # ISO variants across GNU and BSD date.
+          iso_to_epoch() {
+            local iso="$1"
+            # GNU date: accepts the ISO string as-is (with or without Z).
+            date -u -d "$iso" +%s 2>/dev/null && return 0
+            # GNU date: append Z for naive PyPI upload_time.
+            date -u -d "${iso}Z" +%s 2>/dev/null && return 0
+            # BSD date (macOS): strip trailing Z and parse via -jf.
+            local stripped="${iso%Z}"
+            date -u -jf '%Y-%m-%dT%H:%M:%S' "$stripped" +%s 2>/dev/null && return 0
+            return 1
+          }
+
+          # fetch_github <owner> <repo> <version> — print published_at ISO on stdout, return 1 on failure.
+          fetch_github() {
+            local owner="$1" repo="$2" version="$3"
+            if [ -n "${AGE_FIXTURE_DIR:-}" ]; then
+              local fx="$AGE_FIXTURE_DIR/github/$owner/$repo/releases/tags/v$version.json"
+              [ -f "$fx" ] || return 1
+              jq -r '.published_at // empty' "$fx"
+              return 0
+            fi
+            local resp
+            if ! resp=$(gh api "repos/$owner/$repo/releases/tags/v$version" 2>/dev/null); then
+              sleep 2
+              if ! resp=$(gh api "repos/$owner/$repo/releases/tags/v$version" 2>/dev/null); then
+                return 1
+              fi
+            fi
+            printf '%s' "$resp" | jq -r '.published_at // empty'
+          }
+
+          # fetch_pypi <pkg> <version> — print "<upload_time>\t<yanked_bool>" on stdout, return 1 on failure.
+          fetch_pypi() {
+            local pkg="$1" version="$2"
+            local upload yanked
+            if [ -n "${AGE_FIXTURE_DIR:-}" ]; then
+              local fx="$AGE_FIXTURE_DIR/pypi/$pkg/$version.json"
+              [ -f "$fx" ] || return 1
+              upload=$(jq -r '.urls[0].upload_time // empty' "$fx")
+              yanked=$(jq -r '.urls[0].yanked // false' "$fx")
+              printf '%s\t%s\n' "$upload" "$yanked"
+              return 0
+            fi
+            local resp
+            if ! resp=$(curl -sf --max-time 30 "https://pypi.org/pypi/$pkg/$version/json" 2>/dev/null); then
+              sleep 2
+              if ! resp=$(curl -sf --max-time 30 "https://pypi.org/pypi/$pkg/$version/json" 2>/dev/null); then
+                return 1
+              fi
+            fi
+            upload=$(printf '%s' "$resp" | jq -r '.urls[0].upload_time // empty')
+            yanked=$(printf '%s' "$resp" | jq -r '.urls[0].yanked // false')
+            printf '%s\t%s\n' "$upload" "$yanked"
+          }
+
+          # emit name version ecosystem published age verdict reason
+          emit() {
+            printf '%s\t%s\t%s\t%s\t%s\t%s\t%s\n' "$1" "$2" "$3" "$4" "$5" "$6" "$7"
+          }
+
+          while IFS=$'\t' read -r name version ecosystem || [ -n "${name:-}" ]; do
+            [ -z "${name:-}" ] && continue
+
+            # Escape hatch: COOLDOWN_DAYS=0 → all pass without lookup.
+            if [ "$COOLDOWN_DAYS" -eq 0 ]; then
+              emit "$name" "$version" "$ecosystem" "-" "-" "pass" ""
+              continue
+            fi
+
+            case "$ecosystem" in
+              actions)
+                owner="${name%%/*}"
+                remainder="${name#*/}"
+                repo="${remainder%%/*}"
+                if ! iso=$(fetch_github "$owner" "$repo" "$version"); then
+                  emit "$name" "$version" "$ecosystem" "-" "-" "error" "tier-1-404"
+                  continue
+                fi
+                if [ -z "$iso" ]; then
+                  emit "$name" "$version" "$ecosystem" "-" "-" "error" "transient-failure"
+                  continue
+                fi
+                if ! pub_epoch=$(iso_to_epoch "$iso"); then
+                  emit "$name" "$version" "$ecosystem" "-" "-" "error" "parse-failure"
+                  continue
+                fi
+                age_days=$(( (NOW_EPOCH - pub_epoch) / 86400 ))
+                if [ "$age_days" -ge "$COOLDOWN_DAYS" ]; then
+                  emit "$name" "$version" "$ecosystem" "$iso" "$age_days" "pass" ""
+                else
+                  emit "$name" "$version" "$ecosystem" "$iso" "$age_days" "fail" ""
+                fi
+                ;;
+
+              pypi)
+                if ! result=$(fetch_pypi "$name" "$version"); then
+                  emit "$name" "$version" "$ecosystem" "-" "-" "error" "pypi-404"
+                  continue
+                fi
+                iso="${result%%$'\t'*}"
+                yanked="${result##*$'\t'}"
+                if [ -z "$iso" ]; then
+                  emit "$name" "$version" "$ecosystem" "-" "-" "error" "transient-failure"
+                  continue
+                fi
+                if ! pub_epoch=$(iso_to_epoch "$iso"); then
+                  emit "$name" "$version" "$ecosystem" "-" "-" "error" "parse-failure"
+                  continue
+                fi
+                age_days=$(( (NOW_EPOCH - pub_epoch) / 86400 ))
+                if [ "$yanked" = "true" ]; then
+                  emit "$name" "$version" "$ecosystem" "$iso" "$age_days" "fail" "yanked"
+                elif [ "$age_days" -ge "$COOLDOWN_DAYS" ]; then
+                  emit "$name" "$version" "$ecosystem" "$iso" "$age_days" "pass" ""
+                else
+                  emit "$name" "$version" "$ecosystem" "$iso" "$age_days" "fail" ""
+                fi
+                ;;
+
+              *)
+                emit "$name" "$version" "$ecosystem" "-" "-" "error" "unknown-ecosystem"
+                ;;
+            esac
+          done
+
+          exit 0
+          )
+          # --- END inline:scripts/check-release-age.sh ---
+
           DIFF=$(gh pr diff "$PR_NUMBER" --repo "$GH_REPO")
 
           DEPS=""
@@ -235,7 +385,7 @@ jobs:
           COOLDOWN_FAILURES=0
           AGE_TSV=""
           if [ "$COOLDOWN_DAYS" -gt 0 ]; then
-            AGE_TSV=$(echo "$DEPS_TSV" | bash "${GITHUB_WORKSPACE}/shared-workflows/scripts/check-release-age.sh")
+            AGE_TSV=$(echo "$DEPS_TSV" | check_release_age)
             COOLDOWN_FAILURES=$(echo "$AGE_TSV" | awk -F'\t' '$6=="fail" || $6=="error"' | wc -l | tr -d ' ')
           fi
 

--- a/.github/workflows/dependency-cooldown.yml
+++ b/.github/workflows/dependency-cooldown.yml
@@ -23,20 +23,17 @@ on:
 jobs:
   scan:
     runs-on: ubuntu-latest
+    # Gate discipline (see #29): `Set initial status` must be the first
+    # step after `Harden runner`. Every subsequent step must carry
+    # `if: steps.gate.outputs.skip == 'false'` unless it is itself part
+    # of the gate. This avoids making non-bot PRs pay for fallible work
+    # (cross-repo fetches, API calls, script execution) that would only
+    # be used for Dependabot PRs.
     steps:
       - name: Harden runner
         uses: step-security/harden-runner@fa2e9d605c4eeb9fcad4c99c224cee0c6c7f3594 # v2.16.0
         with:
           egress-policy: audit
-
-      - name: Checkout shared-workflows scripts
-        uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
-        with:
-          repository: j7an/shared-workflows
-          ref: ${{ github.workflow_sha }}
-          path: shared-workflows
-          sparse-checkout: scripts
-          persist-credentials: false
 
       - name: Set initial status
         id: gate

--- a/scripts/check-inline-sync.sh
+++ b/scripts/check-inline-sync.sh
@@ -1,0 +1,61 @@
+#!/usr/bin/env bash
+# check-inline-sync.sh — verify inline bash in dependency-cooldown.yml
+# matches scripts/*.sh byte-for-byte after the known normalization rules
+# (YAML indent strip, function wrapper strip, shebang strip).
+#
+# Usage: ./scripts/check-inline-sync.sh
+# Exit:  0 on success, 1 on any drift, missing sentinel, or residual
+#        runtime checkout reference.
+
+set -euo pipefail
+
+WORKFLOW=".github/workflows/dependency-cooldown.yml"
+SCRIPTS=(
+  "scripts/extract-deps.sh"
+  "scripts/check-release-age.sh"
+)
+YAML_INDENT="          "  # exactly 10 spaces — matches the `run: |` indent
+
+fail=0
+
+extract_inline_body() {
+  local script_path="$1"
+  local begin_marker="# --- BEGIN inline:${script_path} ---"
+  local end_marker="# --- END inline:${script_path} ---"
+
+  awk -v begin="$begin_marker" -v end="$end_marker" '
+    index($0, begin) { inside=1; next }
+    index($0, end)   { inside=0; next }
+    inside { print }
+  ' "$WORKFLOW"
+}
+
+for script in "${SCRIPTS[@]}"; do
+  if ! grep -qF "# --- BEGIN inline:${script} ---" "$WORKFLOW"; then
+    echo "FAIL: no inline sentinel for '${script}' found in ${WORKFLOW}"
+    fail=1
+    continue
+  fi
+
+  # Strip first and last lines (the `fn() (` opener and the `)` closer),
+  # then strip the 10-space YAML indent from every remaining line.
+  inline_body=$(extract_inline_body "$script" | sed -E '1d;$d' | sed -E "s/^${YAML_INDENT}//")
+  standalone_body=$(tail -n +2 "$script")
+
+  if ! diff <(printf '%s\n' "$inline_body") <(printf '%s\n' "$standalone_body") > /dev/null 2>&1; then
+    echo "FAIL: inline copy of '${script}' in ${WORKFLOW} does not match source:"
+    diff <(printf '%s\n' "$inline_body") <(printf '%s\n' "$standalone_body") || true
+    fail=1
+  else
+    echo "OK:   ${script} inline copy matches source"
+  fi
+done
+
+# Safety net: residual runtime source-fetch references must not exist.
+if grep -qF '${GITHUB_WORKSPACE}/shared-workflows/scripts/' "$WORKFLOW"; then
+  echo "FAIL: residual runtime source-fetch reference in ${WORKFLOW}:"
+  grep -nF '${GITHUB_WORKSPACE}/shared-workflows/scripts/' "$WORKFLOW" || true
+  fail=1
+fi
+
+exit $fail

